src/app.rs

```diff
@@ -317,23 +326,109 @@ impl App {
+            .await?;
 
         Ok(())
     }
 
-    fn merge_env_provers(
-        options: batch_insertion::Options,
-        existing_provers: &mut Provers,
-    ) -> Provers {
-        let options_set: HashSet<DbProverConf> = options
+    /// Queues a deletion from the merkle tree.
+    ///
+    /// # Errors
+    ///
+    /// Will return `Err` if identity is already queued, not in the tree, or the
+    /// queue malfunctions.
+    #[instrument(level = "debug", skip(self))]
+    pub async fn delete_identity(&self, commitment: &Hash) -> Result<(), ServerError> {
+        // Ensure that deletion provers exist
+        if !self.identity_manager.has_deletion_provers().await {
+            warn!(
+                ?commitment,
+                "Identity Manager has no deletion provers. Add provers with /addBatchSize request."
+            );
+            return Err(ServerError::NoProversOnIdDeletion);
+        }
+
+        // Get the leaf index for the id commitment
+        let leaf_index = self
+            .database
+            .get_identity_leaf_index(&commitment)
+            .await?
+            .ok_or(ServerError::IdentityCommitmentNotFound)?
+            .leaf_index;
+
+        // Check if the id has already been deleted
+        if self.tree_state.get_latest_tree().get_leaf(leaf_index) == Uint::ZERO {
+            return Err(ServerError::IdentityAlreadyDeleted);
+        }
+
+        // Check if the id is already queued for deletion
+        if self
+            .database
+            .identity_is_queued_for_deletion(commitment)
+            .await?
+        {
+            return Err(ServerError::IdentityQueuedForDeletion);
+        }
+
+        // If the id has not been deleted, insert into the deletions table
+        self.database
+            .insert_new_deletion(leaf_index, &commitment)
+            .await?;
+
+        Ok(())
+    }
+
+    /// Queues a deletion from the merkle tree.
+    ///
+    /// # Errors
+    ///
+    /// Will return `Err` if identity is already queued, not in the tree, or the
+    /// queue malfunctions.
+    #[instrument(level = "debug", skip(self))]
+    pub async fn recover_identity(
+        &self,
+        existing_commitment: &Hash,
+        new_commitment: &Hash,
+    ) -> Result<(), ServerError> {
+        // Ensure that insertion provers exist
+        if !self.identity_manager.has_insertion_provers().await {
+            warn!(
+                ?new_commitment,
+                "Identity Manager has no provers. Add provers with /addBatchSize request."
+            );
+            return Err(ServerError::NoProversOnIdInsert);
+        }
+
+        // Ensure that deletion provers exist
+        if !self.identity_manager.has_deletion_provers().await {
+            warn!(
+                ?new_commitment,
+                "Identity Manager has no deletion provers. Add provers with /addBatchSize request."
+            );
+            return Err(ServerError::NoProversOnIdDeletion);
+        }
```

Comment:
That's redundant because of `self.delete_identity` no?

Reply:
Yes, good catch. Removing this snippet now.